File: mobile/shared/src/commonTest/kotlin/uk/foodclassifier/IngredientChipsTest.kt
package uk.foodclassifier

import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class IngredientChipsTest {
    @Test
    fun parsesGarbledCameraOcrIntoMultipleChips() {
        val garbled =
            "B-NMGREDIENTS:Protein Blend (Calcium Caseinate (MUK, NL-IM Whey Protein Isolate Milk;; " +
                "Sweeteners: Maltitol, Sucralose; Bovine Collagen Hydrolysate;; " +
                "Humectant: Glycerol; Water, PRln, Cocoa Butter, Whole Milk Powder;; " +
                "Emulsifier Soy Lecithin; Panuings. ALLERGN VICE,"

        val chips = parseIngredientsToChips(garbled)

        assertTrue(chips.size > 5, "Expected multiple chips but got: $chips")
        assertTrue(chips.any { it.contains("Maltitol", ignoreCase = true) })
        assertTrue(chips.any { it.contains("Glycerol", ignoreCase = true) })
        assertTrue(chips.any { it.contains("Collagen", ignoreCase = true) })
        assertFalse(chips.any { it.contains("B-NMGREDIENTS", ignoreCase = true) })
    }

    @Test
    fun parsesMultilineExtractedIngredientsIntoChips() {
        val chips = parseIngredientsToChips(
            """
            Reduced Fat Soft Cheese (Milk),
            Salt,
            Stabilisers: Guar Gum;
            Carrageenan,
            """.trimIndent(),
        )
        assertEquals(4, chips.size)
        assertEquals("Reduced Fat Soft Cheese (Milk)", chips[0])
        assertEquals("Guar Gum", chips[2])
    }

    @Test
    fun stripsFunctionalLabelsFromChips() {
        val chips = parseIngredientsToChips(
            "Sweeteners: Maltitol, Sucralose, Emulsifier: Soy Lecithin, Acidity Regulator: Sodium Hydroxide",
        )
        assertEquals(
            listOf("Maltitol", "Sucralose", "Soy Lecithin", "Sodium Hydroxide"),
            chips,
        )
    }

    @Test
    fun formatsChipsForClassification() {
        assertEquals(
            "water, oats, salt",
            formatIngredientsFromChips(listOf("water", "oats", "salt")),
        )
    }
}

File: flutter/test/ingredient_chips_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:food_classifier/food_classifier.dart';

void main() {
  test('parsesGarbledCameraOcrIntoMultipleChips', () {
    const garbled =
        'B-NMGREDIENTS:Protein Blend (Calcium Caseinate (MUK, NL-IM Whey Protein Isolate Milk;; '
        'Sweeteners: Maltitol, Sucralose; Bovine Collagen Hydrolysate;; '
        'Humectant: Glycerol; Water, PRln, Cocoa Butter, Whole Milk Powder;; '
        'Emulsifier Soy Lecithin; Panuings. ALLERGN VICE,';

    final chips = parseIngredientsToChips(garbled);

    expect(chips.length > 5, isTrue, reason: 'Expected multiple chips but got: $chips');
    expect(chips.any((chip) => chip.toLowerCase().contains('maltitol')), isTrue);
    expect(chips.any((chip) => chip.toLowerCase().contains('glycerol')), isTrue);
    expect(chips.any((chip) => chip.toLowerCase().contains('collagen')), isTrue);
    expect(
      chips.any((chip) => chip.toLowerCase().contains('b-nmgredients')),
      isFalse,
    );
  });

  test('parsesMultilineExtractedIngredientsIntoChips', () {
    final chips = parseIngredientsToChips('''
Reduced Fat Soft Cheese (Milk),
Salt,
Stabilisers: Guar Gum;
Carrageenan,
''');
    expect(chips, hasLength(4));
    expect(chips[0], 'Reduced Fat Soft Cheese (Milk)');
    expect(chips[2], 'Guar Gum');
  });

  test('stripsFunctionalLabelsFromChips', () {
    final chips = parseIngredientsToChips(
      'Sweeteners: Maltitol, Sucralose, Emulsifier: Soy Lecithin, Acidity Regulator: Sodium Hydroxide',
    );
    expect(
      chips,
      ['Maltitol', 'Sucralose', 'Soy Lecithin', 'Sodium Hydroxide'],
    );
  });

  test('formatsChipsForClassification', () {
    expect(
      formatIngredientsFromChips(['water', 'oats', 'salt']),
      'water, oats, salt',
    );
  });
}
